### PR TITLE
HPCC-28608 LDAP Scope Delete on child scopes fails without message

### PR DESCRIPTION
**Pull request by RussWhitehead**
Throw error if deleting a scope fails, providing the user with a meaningful error message. Also slipped in additional logging of LDAP Config

Signed-off-by: Russ Whitehead <russwhitehead@yahoo.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by wangkx — COMMENTED** [review]
Is this the correct way to notify the caller about the error? When the caller requests to delete multiple multiple resources or delete a user with multiple resources, should we stop the whole process in the middle? Or we may collect the error(s) into a buffer and throw them at the end of the process?

File: esp/services/ws_access/ws_accessService.cpp
Comment (wangkx) [inline esp/services/ws_access/ws_accessService.cpp:2096]: Minor: The newline should be removed.

**Review by wangkx — APPROVED** [review]
@RussWhitehead looks fine now.

**Comment by RussWhitehead** [comment]
@richardkchapman @ghalliday Please merge

File: esp/services/ws_access/ws_accessService.cpp
Comment (ghalliday) [inline esp/services/ws_access/ws_accessService.cpp:2117]: @afishbeck  Is this the correct way to return an exception?  From what I can see the callers catch exceptions that are thrown, and then serialize it in different ways depending on the request format.

Comment (RussWhitehead) [reply esp/services/ws_access/ws_accessService.cpp:2117]: @afishbeck @ghalliday The reason it was done this way is that I don't want to abort the operation if any single "delete" fails, instead I want to proceed with the remaining ones and report back all the ones that failed

Comment (ghalliday) [reply esp/services/ws_access/ws_accessService.cpp:2117]: yes, that makes sense.  The question was more whether this code here, once the failures have been collected should be thrown as an exception.  (It is a question from ignorance since I don't know what the general pattern is for these services.)

Comment (RussWhitehead) [reply esp/services/ws_access/ws_accessService.cpp:2117]: @ghalliday @afishbeck It used to throw, so I took your recommendation and, if there are errors,  throw an exception after all deletes are attempted

**Review by ghalliday — APPROVED** [review]
Looks good.  Will merge once @afishbeck approves

File: esp/services/ws_access/ws_accessService.cpp
Comment (afishbeck) [reply esp/services/ws_access/ws_accessService.cpp:2117]: @RussWhitehead @ghalliday sorry it just occurred to me that the exception being thrown when there are multiple is an IMultiException.  In the same place where you catch the exception and append to your error string you should be appending to an IMultiException.  Then throwing that at the end.

The exceptions returned in the response is an array.

File: esp/services/ws_access/ws_accessService.cpp
Comment (afishbeck) [inline esp/services/ws_access/ws_accessService.cpp:2101]: I would also consider when there are more than one exceptions having a new return code that means "multiple errors".  Then the caller will know they can iterate through them all.

**Review by afishbeck — CHANGES_REQUESTED** [review]

**Review by afishbeck — CHANGES_REQUESTED** [review]
Some suggested changes regarding throwing multiple exceptions.

File: esp/services/ws_access/ws_accessService.cpp
Comment (afishbeck) [reply esp/services/ws_access/ws_accessService.cpp:2117]: Here's a fairly simple example:

https://github.com/hpcc-systems/HPCC-Platform/blob/master/esp/services/ws_workunits/ws_workunitsService.cpp#L882

File: esp/services/ws_access/ws_accessService.cpp
Comment (afishbeck) [inline esp/services/ws_access/ws_accessService.cpp:2098]: The change isn't strictly backward compatible if a client is coded to look for this retcode and msg, but discussed with Russ and he thinks it's ok.  This function could return exceptions before, just not in this case, so clients should look for them anyway.

**Review by afishbeck — APPROVED** [review]
Extra newline, and discussion about backward compatibility but overall looks fine.

**Comment by RussWhitehead** [comment]
@ghalliday Please merge

File: esp/services/ws_access/ws_accessService.cpp
Comment (ghalliday) [inline esp/services/ws_access/ws_accessService.cpp:2110]: trivial:  throw me.getClear() is slightly more efficient.  Does not matter in this context.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**
